### PR TITLE
Disambiguate types with same name from different namespaces in mapToTypeNodes

### LINKED ISSUE #33696: Disambiguate union errors when type name is the same, but namespace differs
## Search Terms

- namespaces error
- same type error
- ambiguous type name error
- namespace type name error
- union type same name
- "does not satisfy the constraint" duplicate
- "does not satisfy the constraint" error 
- "is not assignable to type" duplicate
- "is not assignable to type" duplicate error
- "is not assignable to type" namespace error 

Found nothing, but my issue title turned this up: https://github.com/microsoft/TypeScript/issues/1419, which appears to be not-entirely-unrelated, but has been marked as "fixed".

## Suggestion

When multiple variants of the same union have the same type name, but are in different namespaces, assignability errors repeat the type name without qualification. The following code reproduces the issue:

```typescript
namespace Foo {
  export type Yep = { type: "foo.yep" };
}

namespace Bar {
  export type Yep = { type: "bar.yep" };
}

const x = { type: "wat.nup" };
const val: Foo.Yep | Bar.Yep = x;
```

The following error is produced using 3.6.3:

```
error TS2322: Type '{ type: string; }' is not assignable to type 'Yep | Yep'.
  Type '{ type: string; }' is not assignable to type 'Yep'.
    Types of property 'type' are incompatible.
      Type 'string' is not assignable to type '"bar.yep"'.
```

I think it would help a lot if the error message was able to include as many parent namespace segments as necessary to present the type names unambiguously:

```
error TS2322: Type '{ type: string; }' is not assignable to type 'Foo.Yep | Bar.Yep'.
```

## Use Cases

We use namespaces extensively in our codebase, often in unions with many more variants than the above example. It's impossible to debug typing issues when we are confronted with error messages like this:

    Type '...' is not assignable to type 'Entity | Entity | Entity | Entity | Entity | Entity | undefined'

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

### PR DESCRIPTION
**Pull request by andrewbranch**
Fixes #33696

### CODE REVIEW

**Comment by andrewbranch** [comment]
@typescript-bot perf test this
@typescript-bot user test this

**Comment by andrewbranch** [comment]
GitHub Actions seems to be having trouble at the moment

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -4709,13 +4712,42 @@ namespace ts {
...
                             result.push(typeNode);
+                            if (seenNames && isIdentifierTypeReference(typeNode)) {
+                                seenNames.add(typeNode.typeName.escapedText, [type, result.length - 1]);
+                            }
                         }
                     }
 
+                    if (seenNames) {
+                        // To avoid printing types like `[Foo, Foo]` or `Bar & Bar` where
+                        // occurrances of the same name actually come from different
```
Comment (sandersn) [inline src/compiler/checker.ts:4723]:
```suggestion
                        // occurrences of the same name actually come from different

```

File: src/compiler/utilities.ts [hunk 2]
```diff
@@ -6353,4 +6353,18 @@ namespace ts {
         }) as HeritageClause | undefined;
         return heritageClause?.token === SyntaxKind.ImplementsKeyword || heritageClause?.parent.kind === SyntaxKind.InterfaceDeclaration;
     }
+
+    export function isIdentifierTypeReference(node: Node): node is TypeReferenceNode & { typeName: Identifier } {
+        return isTypeReferenceNode(node) && isIdentifier(node.typeName);
+    }
+
+    export function arrayIsHomogeneous<T>(array: readonly T[], comparer: EqualityComparer<T> = equateValues) {
```
Comment (sandersn) [inline src/compiler/utilities.ts:6361]: I would really really like to recommend the obscure haskell name for this function but I can't remember it.

It's probably `eigenall`. See how much shorter that is?

Comment (sandersn) [reply src/compiler/utilities.ts:6361]: Serious comment: I wouldn't make the second parameter have a default until there's a reason for it, unless the default is really easy to get wrong.

Comment (andrewbranch) [reply src/compiler/utilities.ts:6361]: The issue is more that I really had to hunt for `equateValues`, which I think is the expected default for a function by this name. Using it as a default matches the convention set by `contains`, `arraysEqual`, `arrayIsEqualTo`, and `equalOwnProperties` in core.

Comment (sandersn) [reply src/compiler/utilities.ts:6361]: Fair enough; that's what I was getting at with "easy to get the default wrong".

**Comment by sandersn** [comment]
@andrewbranch I think it's probably OK to merge this since @weswigham has had time to review this if he wanted to.

**Pull request merged**